### PR TITLE
add user management commands

### PR DESCRIPTION
**Pull request by mattkim**
## What was changed

add user management commands to `tcld`

Commands added:
- user
  - list: list all users in the account
  - get: get a user in the account
  - invite: invite users to the account along with account role and namespace permissions
  - delete: delete a user from the account
  - resend-invite: resend invite for a user
  - roles
    - list: list all roles for a user
    - set-account-role: set the account role for a user
    - set-namespace-permissions: set namespace permissions for a user

## Why?

give users the ability to automate user management with tcld

## Checklist
1. How was this tested:
  unit tests

### CODE REVIEW

File: app/user.go [hunk 1]
```diff
@@ -0,0 +1,597 @@
...
+		RequestId: ctx.String(RequestIDFlagName),
+	}
+	resp, err := c.client.UpdateUserNamespacePermissions(c.ctx, req)
+	if err != nil {
+		return fmt.Errorf("unable to update user's namespace permission: %w", err)
+	}
+	return PrintProto(resp.GetRequestStatus())
+}
+
+func (c *UserClient) setNamespacePermissions(
```
Comment (mattkim) [inline app/user.go:323]: this doesn't seem like it's actually doing a set :P

Comment (mattkim) [reply app/user.go:323]: i wonder if we should have an `update` and a `set` api.

File: app/role.go [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+	return res, nil
+}
+
+func getNamespaceRole(ctx context.Context, client authservice.AuthServiceClient, namespace string, actionGroup string) (*auth.Role, error) {
+	ag, err := toNamespaceActionGroup(actionGroup)
+	if err != nil {
+		return nil, err
+	}
+	res, err := client.GetRolesByPermissions(ctx, &authservice.GetRolesByPermissionsRequest{
+		Specs: []*auth.RoleSpec{{
```
Comment (mastermanu) [inline app/role.go:55]: can we not do batching and invoke the client less often? Or is that not worth it?

File: app/role.go [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+	}
+	return rv
+}
+
+func getNamespaceRolesFromMap(ctx context.Context, client authservice.AuthServiceClient, namespaceActionGroups map[string]string) ([]*auth.Role, error) {
+	var res []*auth.Role
+	for namespace, actionGroup := range namespaceActionGroups {
+		r, err := getNamespaceRole(ctx, client, namespace, actionGroup)
+		if err != nil {
+			return nil, err
```
Comment (mastermanu) [inline app/role.go:42]: [Nit] is a multierr a better user experience here?

File: app/role.go [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+		}},
+	})
+	if err != nil {
+		return nil, err
+	}
+	if len(res.Roles) == 0 {
+		return nil, fmt.Errorf("no roles found")
+	}
+	if len(res.Roles) > 1 {
+		return nil, fmt.Errorf("more than 1 account role found: %s", res.Roles)
```
Comment (mastermanu) [inline app/role.go:69]: [Nit] typo in error message?

File: app/user.go [hunk 5]
```diff
@@ -0,0 +1,597 @@
...
+		}
+		for _, nsRole := range nsRoles {
+			roleIDs = append(roleIDs, nsRole.GetId())
+		}
+	}
+
+	// Invite users to given roles
+	req := &authservice.InviteUsersRequest{
+		Specs:     make([]*auth.UserSpec, len(emails)),
+		RequestId: ctx.String(RequestIDFlagName),
```
Comment (tminusplus) [inline app/user.go:160]: Should `RequestIDFlagName` be required, since we directly use it in most of these calls?

File: app/user.go [hunk 6]
```diff
@@ -0,0 +1,597 @@
...
+	res, err := c.client.GetUser(c.ctx, &authservice.GetUserRequest{
+		UserEmail: userEmail,
+		UserId:    userID,
+	})
+	if err != nil {
+		return err
+	}
+	req := &authservice.DeleteUserRequest{
+		UserId:          res.User.Id,
+		ResourceVersion: ctx.String(ResourceVersionFlagName),
```
Comment (tminusplus) [inline app/user.go:216]: Should `ResourceVersionFlagName` be required since we use it directly in this command?

File: app/user.go [hunk 7]
```diff
@@ -0,0 +1,597 @@
...
+	return PrintProto(resp.GetRequestStatus())
+}
+
+func (c *UserClient) setAccountRole(
+	ctx *cli.Context,
+	userID string,
+	userEmail string,
+	accountRole string,
+) error {
+	user, userRoles, err := c.getUserAndRoles(userID, userEmail)
```
Comment (tminusplus) [inline app/user.go:252]: Should we make sure one of these is set (or use that MutuallyExclusiveFlag + required that I mentioned earlier)?

File: app/user.go [hunk 8]
```diff
@@ -0,0 +1,597 @@
...
+			}
+			// find the current admin role to replace
+			ar := ur.Spec.AccountRole
+			if ar != nil && ar.ActionGroup != auth.ACCOUNT_ACTION_GROUP_UNSPECIFIED {
+				// only swap the current admin role
+				userRoles[i] = accountRoleToSet
+			}
+		}
+	}
+	roleNames := make([]string, len(userRoles))
```
Comment (tminusplus) [inline app/user.go:287]: nit: might be nicer to allocate capacity `make([]string, 0, len(userRoles))` then append to avoid.

File: app/user.go (hunk 1 above)
Comment (tminusplus) [reply app/user.go:323]: I'm more in favor of calling this `update` and allow the behavior to be decided in the API (upsert or not). I like the AIP guidelines about this, which follows the declarative model we have. https://google.aip.dev/134#create-or-update

File: app/user.go [hunk 9]
```diff
@@ -0,0 +1,597 @@
...
+					},
+				},
+				{
+					Name:    "invite",
+					Usage:   "Invite users to Temporal Cloud",
+					Aliases: []string{"i"},
+					Flags: []cli.Flag{
+						&cli.StringSliceFlag{
+							Name:     userEmailFlagName,
+							Usage:    "The email address of the user, you can supply this flag multiple times to invite multiple users in a single request",
```
Comment (tminusplus) [inline app/user.go:430]: nit: `StringSliceFlag` mentions that it accepts multiple inputs so I think we could shorten this to `The email address(es) of the user to invite`

File: app/user.go [hunk 10]
```diff
@@ -0,0 +1,597 @@
...
+	userEmailFlagName           = "user-email"
+	accountRoleFlagName         = "account-role"
+	namespacePermissionFlagName = "namespace-permission"
+)
+
+var (
+	userIDFlag = &cli.StringFlag{
+		Name:    userIDFlagName,
+		Usage:   "The user id",
+		Aliases: []string{"i"},
```
Comment (mastermanu) [inline app/user.go:25]: [Nit] I think 'id' is better than 'i'

File: app/user.go [hunk 11]
```diff
@@ -0,0 +1,597 @@
...
+						},
+						&cli.StringFlag{
+							Name:     accountRoleFlagName,
+							Usage:    fmt.Sprintf("The account role to set on the user; valid types are: %v", accountActionGroups),
+							Aliases:  []string{"ar"},
+							Required: true,
+						},
+						&cli.StringSliceFlag{
+							Name:    namespacePermissionFlagName,
+							Usage:   fmt.Sprintf("Flag can be used multiple times; value must be \"namespace=permission\"; valid types are: %v", namespaceActionGroups),
```
Comment (tminusplus) [inline app/user.go:442]: Could we use a [StringMapFlag](https://pkg.go.dev/github.com/urfave/cli/v3#StringMapFlag) instead? It will enforce the format `namespace=permission`.

File: app/user.go [hunk 12]
```diff
@@ -0,0 +1,597 @@
...
+	"github.com/temporalio/tcld/protogen/api/authservice/v1"
+	"github.com/urfave/cli/v2"
+	"strings"
+)
+
+const (
+	userIDFlagName              = "user-id"
+	userEmailFlagName           = "user-email"
+	accountRoleFlagName         = "account-role"
+	namespacePermissionFlagName = "namespace-permission"
```
Comment (mastermanu) [inline app/user.go:18]: I think we added a --user-namespace-permission flag for namespaces? Do we want to try and consolidate?

File: app/user.go [hunk 13]
```diff
@@ -0,0 +1,597 @@
...
+
+var (
+	userIDFlag = &cli.StringFlag{
+		Name:    userIDFlagName,
+		Usage:   "The user id",
+		Aliases: []string{"i"},
+	}
+	userEmailFlag = &cli.StringFlag{
+		Name:    userEmailFlagName,
+		Usage:   "The user email address of the user to be invited",
```
Comment (mastermanu) [inline app/user.go:29]: is the usage of this flag guaranteed to ALWAYS be in the context of invitations?

File: app/user.go [hunk 14]
```diff
@@ -0,0 +1,597 @@
...
+							Usage:   "Delete namespace permissions for a user",
+							Aliases: []string{"dnp"},
+							Flags: []cli.Flag{
+								userIDFlag,
+								userEmailFlag,
+								RequestIDFlag,
+								ResourceVersionFlag,
+								&cli.StringSliceFlag{
+									Name:     NamespaceFlagName,
+									Usage:    "The namespace to delete user permissions from",
```
Comment (tminusplus) [inline app/user.go:578]: nit:  `The namespace to delete` -> `The namespace(s) to delete`

File: app/user.go [hunk 15]
```diff
@@ -0,0 +1,597 @@
...
+							Aliases:  []string{"e"},
+							Required: true,
+						},
+						&cli.StringFlag{
+							Name:     accountRoleFlagName,
+							Usage:    fmt.Sprintf("The account role to set on the user; valid types are: %v", accountActionGroups),
+							Aliases:  []string{"ar"},
+							Required: true,
+						},
+						&cli.StringSliceFlag{
```
Comment (tminusplus) [inline app/user.go:440]: Could be nice to define this once since we use it in multiple places.

File: app/user.go (hunk 11 above)
Comment (tminusplus) [inline app/user.go:442]: nit: the help already mentions it accepts multiple inputs, could be nice to describe what it does `sets the user's permission for the namespace`.

File: app/role.go [hunk 16]
```diff
@@ -0,0 +1,128 @@
...
+	}
+	return res.Roles[0], nil
+}
+
+func toAccountActionGroup(actionGroup string) (auth.AccountActionGroup, error) {
```
Comment (tminusplus) [inline app/role.go:98]: nit: I think on all the action group flags, we could add a `SuggestFlagFunc` for them which provides the accepted values. Should make validation easier and I think this enables auto-complete in user's terminals when using our cli.

**Review by tminusplus — APPROVED** [review]
Mostly style / small comments. I'm happy to come back and address some of the comments later if we need to get this in sooner.

File: app/user.go [hunk 17]
```diff
@@ -0,0 +1,597 @@
...
+		Aliases: []string{"i"},
+	}
+	userEmailFlag = &cli.StringFlag{
+		Name:    userEmailFlagName,
+		Usage:   "The user email address of the user to be invited",
+		Aliases: []string{"e"},
+	}
+)
+
+type UserClient struct {
```
Comment (mastermanu) [inline app/user.go:34]: do we follow this same pattern for other resource types? Also, why is this exported

File: app/user.go [hunk 18]
```diff
@@ -0,0 +1,597 @@
...
+		}
+	}
+}
+
+func (c *UserClient) getUser(userID, userEmail string) (*auth.User, error) {
+	if userID == "" && userEmail == "" {
+		return nil, fmt.Errorf("both user-id and user-email not set")
+	}
+
+	if userID != "" && userEmail != "" {
```
[the remark this answers is not in the document]
Comment (mastermanu) [reply app/user.go:79]: this is pretty neat, but we are currently using v2, and it seems the v3 package is alpha. Does this exist in v2?

File: app/user.go [hunk 19]
```diff
@@ -0,0 +1,597 @@
...
+		pageToken = res.NextPageToken
+		if len(pageToken) == 0 {
+			return PrintProto(totalRes)
+		}
+	}
+}
+
+func (c *UserClient) getUser(userID, userEmail string) (*auth.User, error) {
+	if userID == "" && userEmail == "" {
+		return nil, fmt.Errorf("both user-id and user-email not set")
```
Comment (mastermanu) [inline app/user.go:76]: [Nit] error message implies that user should set both. we should say exactly one must be set

File: app/user.go [hunk 20]
```diff
@@ -0,0 +1,597 @@
...
+		return errors.New("account role required for inviting new users")
+	}
+
+	// the role ids to invite the users for
+	var roleIDs []string
+
+	// first get the required account role
+	role, err := getAccountRole(c.ctx, c.client, accountRole)
+	if err != nil {
+		return err
```
Comment (mastermanu) [inline app/user.go:138]: is the error returned from this clear to someone trying to invite a user? Should we wrap the error message here with "invalid account role specified: %v" ?

File: app/user.go (hunk 5 above)
Comment (mastermanu) [reply app/user.go:160]: the server assigns one automatically if its not supplied by creating a new UUID. What we could do if its not specified is generate the UUID ourselves maybe?

File: app/user.go [hunk 21]
```diff
@@ -0,0 +1,597 @@
...
+	}
+	return PrintProto(resp.GetRequestStatus())
+}
+
+func (c *UserClient) deleteUser(
+	ctx *cli.Context,
+	userID string,
+	userEmail string,
+) error {
+	if userID != "" && userEmail != "" {
```
Comment (mastermanu) [inline app/user.go:200]: can we just call getUser() to avoid duplicating this code?

File: app/user.go (hunk 6 above)
Comment (mastermanu) [reply app/user.go:216]: its optional given we populate it from the existing user if the user doesn't supply a specific version

File: app/user.go (hunk 7 above)
Comment (mastermanu) [reply app/user.go:252]: getUserAndRoles is doing that check underneath the covers, right?

File: app/user.go [hunk 22]
```diff
@@ -0,0 +1,597 @@
...
+		if userRoles[i].Id == accountRoleToSet.Id {
+			return fmt.Errorf("user already has '%s' account role", accountRoleToSet.Id)
+		}
+	}
+	// check if this is the global admin role, and replace all existing roles
+	if accountRoleToSet.Spec.AccountRole.ActionGroup == auth.ACCOUNT_ACTION_GROUP_ADMIN {
+		// set the user account admin role
+		y, err := ConfirmPrompt(ctx, "Setting admin role on user, please confirm")
+		if err != nil || !y {
+			return err
```
Comment (mastermanu) [inline app/user.go:270]: if err == nil and y = false, we are going to return a nil error? That's probably fine, but then maybe we should print out "operation canceled" or something like that?

File: app/user.go [hunk 23]
```diff
@@ -0,0 +1,597 @@
...
+	}
+	for i := range userRoles {
+		if userRoles[i].Id == accountRoleToSet.Id {
+			return fmt.Errorf("user already has '%s' account role", accountRoleToSet.Id)
+		}
+	}
+	// check if this is the global admin role, and replace all existing roles
+	if accountRoleToSet.Spec.AccountRole.ActionGroup == auth.ACCOUNT_ACTION_GROUP_ADMIN {
+		// set the user account admin role
+		y, err := ConfirmPrompt(ctx, "Setting admin role on user, please confirm")
```
Comment (mastermanu) [inline app/user.go:268]: should the message should indicate all existing namespace permissions will be replaced?

File: app/user.go [hunk 24]
```diff
@@ -0,0 +1,597 @@
...
+		if err != nil || !y {
+			return err
+		}
+		userRoles = []*auth.Role{accountRoleToSet}
+	} else {
+		for i := range userRoles {
+			ur := userRoles[i]
+			if ur.Type != auth.ROLE_TYPE_PREDEFINED {
```
Comment (mastermanu) [inline app/user.go:276]: can you remind me what this check is for?

File: app/user.go [hunk 25]
```diff
@@ -0,0 +1,597 @@
...
+				continue
+			}
+			// find the current admin role to replace
+			ar := ur.Spec.AccountRole
+			if ar != nil && ar.ActionGroup != auth.ACCOUNT_ACTION_GROUP_UNSPECIFIED {
+				// only swap the current admin role
+				userRoles[i] = accountRoleToSet
```
Comment (mastermanu) [inline app/user.go:283]: should we break after this assignment?
also is there any scenario where we exit the for loop without doing any update at all? should we error if that happens?

Comment (mastermanu) [reply app/user.go:283]: also whats the 2nd check for? in case there is a future admin role we add that the CLI doesn't understand? But in that case, then CLI cannot be used to update admin role since it can never replace it?

File: app/user.go [hunk 26]
```diff
@@ -0,0 +1,597 @@
...
+	userEmail string,
+	accountRole string,
+) error {
+	user, userRoles, err := c.getUserAndRoles(userID, userEmail)
+	if err != nil {
+		return err
+	}
+	accountRoleToSet, err := getAccountRole(c.ctx, c.client, accountRole)
+	if err != nil {
+		return err
```
Comment (mastermanu) [inline app/user.go:258]: same question, it's not obvious if an error returned from the underlying call has enough context to be friendly to the user

File: app/user.go [hunk 27]
```diff
@@ -0,0 +1,597 @@
...
+	npm, err := toNamespacePermissionsMap(namespacePermissions)
+	if err != nil {
+		return err
+	}
+	for namespace, permission := range npm {
+		ag, err := toNamespaceActionGroup(permission)
+		if err != nil {
+			return err
+		}
+		if err := c.updateUserNamespacePermissions(ctx, userID, userEmail, namespace, ag); err != nil {
```
Comment (mastermanu) [inline app/user.go:338]: isn't this another case where we can do a batch update here?

Comment (mastermanu) [reply app/user.go:338]: actually, I am pretty sure we HAVE to do this in a single batch

Comment (mastermanu) [reply app/user.go:338]: nevermind, the above statement is wrong. That API can actually be used btw to update multiple permissions either for the same user or across users

File: app/user.go [hunk 28]
```diff
@@ -0,0 +1,597 @@
...
+}
+
+func (c *UserClient) deleteNamespacePermission(
+	ctx *cli.Context,
+	userID string,
+	userEmail string,
+	namespaces []string,
+) error {
+	for _, ns := range namespaces {
+		if err := c.updateUserNamespacePermissions(ctx, userID, userEmail, ns, auth.NAMESPACE_ACTION_GROUP_UNSPECIFIED); err != nil {
```
Comment (mastermanu) [inline app/user.go:352]: wait, our way to update user permissions is to set the action group to unspecified?

File: app/user.go [hunk 29]
```diff
@@ -0,0 +1,597 @@
...
+		Namespace: namespace,
+		UserNamespacePermissions: []*auth.UserNamespacePermissions{
+			{
+				UserId:      user.Id,
+				ActionGroup: actionGroup,
+			},
+		},
+		RequestId: ctx.String(RequestIDFlagName),
+	}
+	resp, err := c.client.UpdateUserNamespacePermissions(c.ctx, req)
```
Comment (mastermanu) [inline app/user.go:316]: does this also check if the user already has the existing permissions? it seems server will fail, so its not really an upsert?

File: app/user.go [hunk 30]
```diff
@@ -0,0 +1,597 @@
...
+	for _, kv := range keyValues {
+		parts := strings.Split(kv, "=")
+		if len(parts) != 2 {
+			return nil, fmt.Errorf("invalid namespace permission \"%s\" must be of format: \"namespace=permission\"", kv)
+		}
+
+		namespace := parts[0]
+		actionGroupValue := parts[1]
+
+		if len(namespace) == 0 {
```
Comment (mastermanu) [inline app/user.go:370]: do we need to check parts[1] length?

**Review by mastermanu — CHANGES_REQUESTED** [review]
left some comments

File: app/print.go [hunk 31]
```diff
@@ -25,15 +26,47 @@ func PrintObj(i interface{}) error {
...
+func PrintProto(m proto.Message) error {
+	ser, err := serializeProto(m)
 	if err != nil {
 		return err
 	}
 	fmt.Printf("%v\n", ser)
 	return nil
 }
+
+func PrintProtoSlice(name string, ms []proto.Message) error {
```
[the remark this answers is not in the document]
Comment (mattkim) [reply app/print.go:50]: Yeah because the slice doesn't support the proto.Message interface.

Interesting not about `gogo/protobuf`--yes we should deprecate it.

File: app/role.go (hunk 3 above)
Comment (mattkim) [reply app/role.go:42]: changed to use batch api so not relevant now.

File: app/role.go (hunk 16 above)
Comment (mattkim) [reply app/role.go:87]: So this is very intriguing--but i couldn't figure out how to get it working and don't see any documentation. Do you have any references?

File: app/user.go (hunk 15 above)
Comment (mattkim) [reply app/user.go:489]: unfortunately not reusable because one flag is required the other it is not.

File: app/user.go (hunk 11 above)
Comment (mattkim) [reply app/user.go:491]: interesting--not available in v2--but something we should look into supporting.

File: app/user.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (tminusplus) [reply app/user.go:88]: Doh you are right. Honestly we have been finding some issues with urfave (if you have an env variable which stores a secret it will print that secret in the help message)... but that is a separate discussion from this PR.

File: app/user.go (hunk 5 above)
Comment (tminusplus) [reply app/user.go:169]: Server assigning one sounds best, that makes sense!

File: app/user.go (hunk 11 above)
Comment (tminusplus) [reply app/user.go:442]: Also doh, searching for urfave go docs brought me to v3.

File: app/role.go (hunk 16 above)
Comment (tminusplus) [reply app/role.go:87]: You know what I tried to look for this when me and @anekkanti discovered that urfave prints env variables into the help message.. and couldn't find anything on it.

File: app/version.go [hunk 32]
```diff
@@ -3,7 +3,7 @@ package app
 import "github.com/urfave/cli/v2"
 
 const (
-	DefaultVersion = "v0.5.0"
+	DefaultVersion = "v0.6.0"
```
Comment (tminusplus) [inline app/version.go:6]: I do wonder if we should cut releases (bump the version number) in a separate PR. It helps keep the commit history clean and easy to understand at a glance.

**Review by tminusplus — APPROVED** [review]
Minor comments otherwise LGTM

File: app/user.go [hunk 33]
```diff
@@ -0,0 +1,560 @@
...
+	"github.com/urfave/cli/v2"
+	"strings"
+)
+
+const (
+	userIDFlagName              = "user-id"
+	userEmailFlagName           = "user-email"
+	accountRoleFlagName         = "account-role"
+	namespacePermissionFlagName = "namespace-permission"
+	permissionFlagName          = "permission"
```
Comment (tminusplus) [inline app/user.go:18]: Is this flag used anywhere? I tried looking and it seems unused.

**Review by mastermanu — APPROVED** [review]

File: app/user.go (hunk 33 above)
Comment (mattkim) [reply app/user.go]: Removed.
